### ISSUE TITLE
Support collision for over 3x3x3 node collision boxes

### ISSUE THREAD

**Issue by VanessaE** [minetest/minetest#2612]
When a collision box is set on a node, and it's set to be larger than 1 node space (for example, to accommodate multi-node-sized meshes), one side of the collision box will be passable/non-colliding.  Consider the following code excerpt from the lrfurn mod in Home Decor:

``` lua
local sofa_cbox = {
    type = "fixed",
    fixed = {-0.5, -0.5, -0.5, 0.5, 0.5, 1.5}
}

minetest.register_node("lrfurn:sofa_"..colour, {
    ......
    selection_box = sofa_cbox,
    collision_box = sofa_cbox,
    ......
})
```

This defines a collision box that stretches forward from the nominal -Z face by 2 meters.  If you walk around to the nominal +Z face (regardless of whether that face points north, west, east, or south), you can walk into the node and all the way through to the -Z end.  Similarly, if the last coordinate is changed from 1.5 to 2.5 (for the 3-node-long version of the sofa that this code came from), you can walk along the length of the collision box for about 2 node lengths, and then suddenly "fall through" into the third node space.

The same thing happens whether you create the collision information via the collision_box field or via the node_box field.

I include the line that sets the _selection_ box here because, like many of my multi-node objects, it uses the same definition as the collision box, but appears to be interpreted correctly - that is pointing at the node always highlights the whole thing exactly as the box definition dictates.

//////////////

EDIT by paramat:
2 node sofas no longer have this problem, it seems to have been a mod code issue that has since been solved, however collision fails for the 3 node long sofas.

**Comment by est31** [comment]
confirmed, the bug seems to depend on which direction you place the sofa in.

**Comment by VanessaE** [comment]
This glitch seems to also affect Homedecor's spiral staircase - the collision data for the top two steps doesn't work, and like the sofa, their coordinates are in the 2.0+ range (but in the Y axis this time).

**Comment by est31** [comment]
This can be "fixed" if the limits for the search in [collision.cpp](https://github.com/minetest/minetest/blob/master/src/collision.cpp#L241-246) are increased, but out of performance reasons, there is a limit.

**Comment by VanessaE** [comment]
[05-30 17:30] <VanessaE> having the engine automatically chop the collision boxes into smaller pieces might help.
[05-30 17:30] <VanessaE> BUT that won't fix e.g the staircase problem
[05-30 17:30] <VanessaE> where its collision data is already cut into many small pieces
[05-30 17:31] <VanessaE> the problem isn't so much the size of the boxes, it's what the engine considers to be the "origin" of the box
[05-30 17:32] <VanessaE> it should auto-adjust internally so that a box's origin is (one of the) node space(s) it actually occupies, rather than the object's "root" node

**Comment by beyondlimits** [comment]
Maybe it is not a bug at all but just the way the engine works. It does not check whether line of sight crosses something that is outside of a nodebox area.

The fact that you can often point something that is outside the nodebox seems to be just a side effect of poorly implemented `getPointedThing` that examines every node in a chunk instead of only those that meet the line.

http://www.academictutorials.com/graphics/graphics-raycasting.asp

You could as well define 100-meter nodebox and claim it's a bug that it doesn't work.

**Comment by VanessaE** [comment]
That might be true in the case of a 100-meter nodebox, but this isn't a nodebox issue, let alone one that size.  This is a collision box issue, as in the nodes in question have `collision_box=` definitions that span no more than three meters for any one collision box entry, and in most cases, less than one meter.

**Comment by paramat** [comment]
Is this high priority? Doesn't seem so.

**Comment by VanessaE** [comment]
No action from the devs for over a year, other than useless labels and categorization.  Closing.

**Comment by paramat** [comment]
Should be kept open as is a significant issue.

**Comment by raymoo** [comment]
You could maintain lists of in each loaded block for each size of node. For example, a list for positions of nodes with max size on one axis > 1 and ≤ 2, a list for nodes with size > 2 and ≤ 3, one for > 3, ≤ 4, etc. Then you can increase the range of included nodes for collision checking in only those lists.

**Comment by paramat** [comment]
That's overkill for these rare situations. Maybe these large objects should be several separate nodes instead.
We certainly cannot increase the range for collision checks as this unnecessarily reduces performance for 99.99% of a world.
I can't see a practical way to detect a nearby large object and increase the collision check radius, and that's a messy and overkill idea anyway.
Seems unsolvable in any reasonable way?

**Comment by raymoo** [comment]
@paramat You would need to keep a separate list:
 - When the block is first read into memory, build the lists of large nodes
 - When the block is modified, check the modified areas for new large nodes or large nodes that were removed

The way I would recommend to check a node's size is to precalculate this for all nodes beforehand and just have an array mapping content IDs to sizes.

**Comment by paramat** [comment]
Yes this is what i mean by overkill. These large objects are very rare and arguably should be implemented differently.

**Comment by raymoo** [comment]
@paramat The alternative implementation I would favor is using entities for collision, but that has the issue that (I think?) Minetest doesn't use an efficient data structure for collision checking entities, and also has a low entities limit.

Using separate nodes is painful to manage, as can be seen with the decision by minetest_game to move to a single door node for collision.

**Comment by raymoo** [comment]
I also don't think it's overkill to maintain the list. Even if you just rebuild the list from scratch every block update, I think Minetest will be able to keep up.

**Comment by paramat** [comment]
> Using separate nodes is painful to manage, as can be seen with the decision by minetest_game to move to a single door node for collision.

The new meshdoors, while good, have caused many issues, we have even considered going back to multiple nodes. Large single nodes cause issues and are not necessarily easier to manage then multilpe-node objects.

I'm sure your udea won't hit performace, my point is that the extreme rarity and dubious implementation of very large nodes doesn't justify the work and maintenance.

**Comment by paramat** [comment]
Actually, thinking back over the issues we had with new doors, i can assure you that the old 2-node doors were certainly less hassle.

I find your suggestion a bit messy and unpleasant, because when a player enters a mapblock that contins large nodes, the collision check range is increased whether or not the player is near or about to travel into those nodes.
If this was high priority then it may be justified, but it's a rare and minor issue, the mod should use multiple nodes, and we already support nodes that extend up to a node beyond the main cube, which includes most large nodes.

**Comment by raymoo** [comment]
@paramat Collision check range would't increased for ordinary nodes, only the large nodes, so the collision check range would be increased exactly for those nodes that need it.

**Comment by paramat** [comment]
Ok, but how would you detect that a player is close enough to a large node to then increase the collision check range? It's still quite complex for a minor issue, and i feel not justified. it's the mod's responsibility to work within the capability of the engine.
This is not a bug, but a chosen limitation of the engine, so this is a feature request.
Sorry but i will disapprove any work on this as being overkill for a very minor issue.

**Comment by raymoo** [comment]
@paramat I don't know what you are talking about, you always check the large nodes with a larger collision check range. Ordinary 1x1x1 nodes will still be checked with the normal range, there is no switching of collision check range.

Here is what I am talking about:
 - There is a space-searchable data structure for 2-sized nodes, 3-sized nodes, etc. (Other schemes are possible, for example a structure for 2-sized, 4-sized, 8-sized, etc.)
 - Nodes for collision check are gathered:
   - All nodes are gathered with the ordinary collision check range (alternatively only the 1-sized nodes, but this is harder to check unless 1-sized nodes are kept track of in addition to the other sizes)
   - The 2-sized nodes are gathered with a collision check range extending 1 more than normal
   - The 3-sized nodes are gathered with a collision check range extending 2 more than normal
   - The n-sized nodes are gathered with a collision check range extending n - 1 more than normal
 - These nodes are then checked for collision.

**Comment by paramat** [comment]
Ok, still overkill for a mod issue and not justified.

**Comment by raymoo** [comment]
@paramat "overkill" means it's more than enough to implement a solution. It's not "overkill", since it's about what's needed to implement it efficiently. To show that it's overkill, you would need to find a solution that's simpler and implements the feature satisfactorily.

**Comment by paramat** [comment]
Ok i agree, what i mean is the work and disruption is not justified, it's a mod issue for a very few nodes.

**Comment by VanessaE** [comment]
This needs re-opened.  I'm sorry @paramat  but you're simply wrong here.

There are many nodes in mods now that are over 1m in size.  The engine's handling here is simply BROKEN.

**Comment by paramat** [comment]
Ok i'm fairly neutral. Sorry for the close that was a mistake, however, you did actually close this yourself twice, and we re-opened it, one of those times i re-opened it.
As explained nothing is broken, we simply support collisions out to a certain distance from a node for performance reasons, so this is a reasonable request to be able to extend this further for large nodes.

As est31 wrote

> This can be "fixed" if the limits for the search in collision.cpp are increased (replace 1 with 2,3,4, etc.), but out of performance reasons, this can't be raised indefinitely. In order to really fix this, we need a way to specify "composite" objects which are larger than 1 node.

**Comment by VanessaE** [comment]
Similarly, my having closed it before was a mistake, both times.

As I stated before, the easiest way to handle this would be for the client to split collision boxes into less-than-1m pieces and alter their "root" coordinates, so if a box runs 1x1x2, split it into two 1x1x1 boxes, with each having coordinates rewritten to act as if the out-past-1m box were simply associated with an invisible node in the correct location.

Basically, have the client convert larger-than-1m boxes into multiple nodes, and have it remove those "dummy" nodes when the root node is removed.

This is best done in-engine to avoid code duplication (else every mod that has these large objects has to add special handling, dummy nodes, etc).

**Comment by paramat** [comment]
It seems to me it's reasonable to support collisionboxes that extend up to 1 node beyond the basic cube, as in your sofa, a 1 node extension is not excessive, further i might consider excessive and low priority. After all, smooth lighting is coded to cope with nodes that extend up to 1 node beyond the basic cube.

**Comment by VanessaE** [comment]
If my idea were used, there would be no limit (nor any need for one).

**Comment by paramat** [comment]
Yeah it sounds like a good idea, although i don't know if it's technically practical.

![screenshot_20170930_073718](https://user-images.githubusercontent.com/3686677/31043389-689f22c6-a5b2-11e7-8168-7ac045e3033d.png)

minetest.register_node("default:obsidian", {
	description = "Obsidian",
	tiles = {"default_obsidian.png"},
	selection_box = {
		type = "fixed",
		fixed = {-0.5, -0.5, -0.5, 0.5, 0.5, 1.5},
	},
	collision_box = {
		type = "fixed",
		fixed = {-0.5, -0.5, -0.5, 0.5, 0.5, 1.5},
	},
	sounds = default.node_sound_stone_defaults(),
	groups = {cracky = 1, level = 2},
})

This has no issues but a collisionbox 3 nodes long does, i'm testing to see at what point the collision fails.

**Comment by paramat** [comment]
```
	selection_box = {
		type = "fixed",
		fixed = {-0.5, -0.5, -0.5, 0.5, 0.5, 1.6},
	},
	collision_box = {
		type = "fixed",
		fixed = {-0.5, -0.5, -0.5, 0.5, 0.5, 1.6},
	},
```
At 1.6 the collision works but the player 'vibrates' against the end (1.6) face.

**Comment by paramat** [comment]
![screenshot_20170930_075906](https://user-images.githubusercontent.com/3686677/31043529-793da08c-a5b5-11e7-9b90-babb39db5cee.png)
```
	selection_box = {
		type = "fixed",
		fixed = {-1.5, -1.5, -1.5, 1.5, 1.5, 1.5},
	},
	collision_box = {
		type = "fixed",
		fixed = {-1.5, -1.5, -1.5, 1.5, 1.5, 1.5},
	},
```
This works too, a 3x3x3 cube. I think anything bigger is quite specialised and not high priority to support. But i'm neutral as long as it does not affect performance in areas with no large nodes, and is not too complex.

**Comment by paramat** [comment]
I think an important issue here is why your shorter sofas have collision issues while being no bigger than my experiment above. if we support a 1 node extension then this should work for your sofa too.
Est31 reports your sofa works in some directions, so perhaps something to do with how it is rotated on place?

**Comment by paramat** [comment]
Tested, your 2 node sofas have no issue but the 3 node sofas do. So it seems we support up to 3x3x3.

**Comment by paramat** [comment]
![screenshot_20170930_091558](https://user-images.githubusercontent.com/3686677/31044084-0a2b4252-a5c0-11e7-9d23-961741b3386b.png)

The spiral staircase is fine up to the step i'm pointing at, then you fall through, i guess because that step is 2 nodes above the base and the base is considerd the 'origin cube'. If somehow the centre of the staricase could be considered the 'origin cube' then it would work.

**Comment by sirrobzeroone** [comment]
I've been reading over this with interest as I may have some meshes that also exceed the 3x3 cube. I understand both angles, sometimes we might as modders want large meshes and an easy way to specify collisions. But also dont want to hammer performance of the base engine.

Given the engine already supports multiple collision_box's inside a 3x3 cube. however those are always specified under the assumption by the engine the origin is 0,0,0. Could an offset be added as a 4(5/6)th optional value which the engine would then move the center point along that axis but leave the size as max 3x3. 

eg
	collision_box = {
		type = "fixed",
		fixed = {-1.5, -1.5, -1.5, 1.5, 1.5, 1.5,1,1,1},
	},

The above would be interrupted by the engine as the center point now being 1,1,1 relative to the mesh center instead of 0,0,0. naturally not specifying would leave the origin at 0,0,0 relative to the mesh.

For a large mesh say  3x6x3 you could simply specify two large 3x3 collision_boxs but one with a y offset of 3 eg

eg
	collision_box = {
		type = "fixed",
		fixed = {-1.5, -1.5, -1.5, 1.5, 1.5, 1.5},
                fixed = {-1.5, -1.5, -1.5, 1.5, 1.5, 1.5, 0, 3, 0},
	},

The above should keep engine speed intact as you still can't specify a mesh beyond 3x3 in size however for large models a modder can simply specify multiple submissions. 

Realize this is similar to Vanessa's solution above but might be less code work for a dev as theres no auto calulating by the engine and onus for getting alignment correct is placed back onto the mod creater. essentially its just three more variables to pass and add/subtract from the current origin calculation.

Anyways just an ideal musing as Im wrestling with some long thin meshes and the above would make my life much easier....yes selfish ends :)

**Comment by SmallJoker** [comment]
@sirrobzeroone The idea is meant well but the collision box would still be bound to the node, hence increasing the node "scan radius". Not the size of the collision box is the issue, but the amount of nodes to check.

**Comment by sirrobzeroone** [comment]
@SmallJoker, ahh thanks that makes sense. That means any engine based workaround is probably not going to be any less cumbersome than a mod based one. I understand paramats comments now. I have come up with a solution mod based for my needs, not sure on performance hit etc yet and im not what you call an elegant coder (amature hack springs to mind). If i can turn it into a standalone mod ill share that uf thats beyond my skill ill just share how ive done it. Im sure Vanessa or one of tge other more skilled modders could tgen do something with it.

**Comment by ExeVirus** [comment]
So I am working on a utility mod/solution right now that actually solves this exact problem, in the exact way you have all suggested: autogenerate multiple nodes that behave as a single "node" for meshes larger than 3x3x3. 

This came from the need to autogenerate collision/selection boxes for meshes and mission creep by myself. I have a working lua-based solution that takes in a given .obj, uses some default quality settings (or user provided), and generates best fit bounding boxes to approximate the mesh with merely boxes. It's relatively efficient (fase) on meshes with under 500 triangles, and anything larger just requires some Meshlab work to simplify back down to under 500 triangles for processing (while still using the original .obj as the viewable version). Here's a screenshot of the results from using an .obj from my HD_rocks mod: 

![image](https://user-images.githubusercontent.com/44562154/96778351-94b99580-13b9-11eb-987e-177c464cb990.png)

As you will notice if you're looking closely, there are two sets of boxes being displayed, and this is by design. The second set is outside the first's 3x3x3 possible collision/selection box area, and so these boxes are separated out for use in "dummy nodes" that hold the collision and selection box data. 

Currently, I am working on encapsulating the data export into something a simple register mod can basically replace minetest.register_node() and will automatically declare and define all the dummy nodes, allow for rotations and placement that actually respects whether it is possible to place those dummy nodes (i.e. there can't be a node already in the middle of a given dummy node spot), and then tells a user when such an action is not possible and why. (should they ask for a verbose why, they simply type /why or similar). 

The end goal is that we can use basically any meshes for nodes, and not kill MT performance trying to do collisions for large nodes. It also does a significantly better job approximating meshes than one can do by hand. For example, I will be replacing HD_Rocks's collision boxes with these boxes as soon as I'm finished, because they're so much more accurate/detailed as I want them. 

The work is part of a larger project called story_npc, but the boxgen stuff can be found here: 
https://github.com/ExeVirus/story_npc/tree/master/boxgen

**Comment by Andrey2470T** [comment]
My suggestion for resolution of this problem is to add a new option adjusted in minetest.conf that would define a maximum distance from the node position in all 6 directions within that to search for nodes for collision. Since there are many mods that have large-sized nodes, default value would be '10.5'.

**Comment by ExeVirus** [comment]
Oh I should have came back here earlier, I did finish my mods to support larger collision boxes (this doesn't work for entities, only nodes with mesh draw type). https://github.com/ExeVirus/Boxgen is complete, and works like a charm, I finished the entire github wiki with pictures and tutorials from beginner to advanced and walkthroughs for working with autogenerating your collision boxes.

I have made the sister mod to autoload these boxes as well: https://github.com/ExeVirus/Autobox
Which supports treating these multibode structures as a single node, it has a single register function, respects protection and other nodes, cleans up on dig from any point, rotates in all directions properly, and supports things you defining your own right click definition, i.e. if you right click and want that to bring up a formspec like a furnace, that works. 


Meanwhile, the reason for the 3x3x3 restriction is mostly how we do collisions and whatnot on the backend, which is mostly accomplished with linear search for many things, I'm currently working on geospatial indexing for entities, which will allow us to do things like collisions, get objects in radius, raycasts, and get objects in area calculations much, much faster. That work is ongoing and will likely take 1-2 more months before the PR is done and another 1-2 months for core devs to approve it, and then however long to make it into official release (maybe 1-2 more months). So roughly fall 2021 this should be implemented and we may be Able to come back and open up collision box sizes. 

In the meantime, go ahead and use my mods, they work well. Autobox comes with a great example of what is possible, and it's quite exciting for others to play with.
